compiler/src/dotty/tools/dotc/core/Substituters.scala

```diff
@@ -195,4 +195,28 @@ trait Substituters { this: Context =>
   final class SubstParamsMap(from: BindingType, to: List[Type]) extends DeepTypeMap {
     def apply(tp: Type): Type = substParams(tp, from, to, this)
   }
+
+  /** An approximating substitution that can handle wildcards in the `to` list */
+  final class SubstApproxMap(from: List[Symbol], to: List[Type])(implicit ctx: Context) extends ApproximatingTypeMap {
+    def apply(tp: Type): Type = tp match {
+      case tp: NamedType =>
+        val sym = tp.symbol
+        var fs = from
+        var ts = to
+        while (fs.nonEmpty) {
+          if (fs.head eq sym)
+            return ts.head match {
+              case TypeBounds(lo, hi) => range(lo, hi)
+              case tp1 => tp1
+            }
+          fs = fs.tail
+          ts = ts.tail
+        }
+        if (tp.prefix `eq` NoPrefix) tp else tp.derivedSelect(apply(tp.prefix))
```

Comment:
I think this should be:
```suggestion
        if (tp.prefix `eq` NoPrefix) tp else derivedSelect(tp, apply(tp.prefix))
```

In fact, maybe we could do this in `def subst` directly, the only difficulty is that `def subst` is supposed to work with `theMap` being `null` too, so we'd need to do something like:
```diff
diff --git compiler/src/dotty/tools/dotc/core/Substituters.scala compiler/src/dotty/tools/dotc/core/Substituters.scala
index ab07a4d568f..c3d999c828d 100644
--- compiler/src/dotty/tools/dotc/core/Substituters.scala
+++ compiler/src/dotty/tools/dotc/core/Substituters.scala
@@ -64,7 +64,8 @@ trait Substituters { this: Context =>
           ts = ts.tail
         }
         if (tp.prefix `eq` NoPrefix) tp
-        else tp.derivedSelect(subst(tp.prefix, from, to, theMap))
+        else if (theMap != null) theMap.derivedSelect(tp, subst(tp.prefix, from, to, theMap))
+        else tp.derivedSelect(tp, subst(tp.prefix, from, to, theMap))
       case _: ThisType | _: BoundType =>
         tp
       case _ =>
```

Then this class could be replaced by something like:
```scala
final def substApprox(tp: Type, from: BindingType, to: BindingType): Type =
  subst(tp, from, to, new ApproximatingTypeMap {
    def apply(tp: Type): Type = subst(tp, from, to, this)
  })
```